SuperChamps/contracts/Metagame/SCMetagamePool.sol

```diff
@@ -0,0 +1,135 @@
+import "../Utils/SCPermissionedAccess.sol";
+import "../../interfaces/IPermissionsManager.sol";
+import "../../interfaces/ISCMetagamePool.sol";
+
+/// @title Metagame staking pool.
+/// @author Chance Santana-Wees (Coelacanth/Coel.eth)
+/// @notice Staking pool for Super Champ tokens.
+/// @dev This pool does not issue on-chain rewards. Rewards are tabulated off-chain.
+contract SCMetagamePool is SCPermissionedAccess, ISCMetagamePool {
+    IERC20 public immutable token;
+
+    mapping(address => mapping(uint256 => uint256)) _user_to_checkpoint_to_balance;
+    mapping(address => uint256[]) _user_checkpoints;
+
+    mapping(address => mapping(address => uint256)) _user_to_approved_spend;
+
+    /// @param permissions_ Address of the protocol permissions registry. Must conform to IPermissionsManager
+    constructor(address permissions_, address token_) SCPermissionedAccess(permissions_) {
+        token = IERC20(token_);
+    }
+
+    /// @notice Stake tokens directly
+    /// @param amount_ Quantity of tokens to stake
+    function stake(uint256 amount_) external {
+        bool success = token.transferFrom(msg.sender, address(this), amount_);
+        if(!success) {
+            revert UnableToTransferTokens(msg.sender, amount_);
+        }
+        _stake(msg.sender, amount_);
+    }
+
+    /// @notice Stake tokens for another address
+    /// @param staker_ The address to stake tokens for
+    /// @param amount_ Quantity of tokens to stake
+    /// @dev Tokens are transferred from msg.sender and credited to staker_
+    function stakeFor(address staker_, uint256 amount_) external {
+        bool success = token.transferFrom(msg.sender, address(this), amount_);
+        if(!success) {
+            revert UnableToTransferTokens(msg.sender, amount_);
+        }
+        _stake(staker_, amount_);
+    }
+
+    /// @notice Approve an address to spend staked tokens
+    /// @param spender_ The address to approve
+    /// @param amount_ Quantity of tokens to approve spending
+    /// @dev This mirrors approve(...) defined in ERC20 standard
+    function approve(address spender_, uint256 amount_) external {
+        _user_to_approved_spend[msg.sender][spender_] += amount_;
+    }
+
+    /// @notice Spend tokens from a stakers staked tokens
+    /// @param amount_ Quantity of tokens to unstake/spend
+    /// @param staker_ The address of the staker
+    /// @param receiver_ The address of the reciever of the spent tokens
+    /// @dev This does NOT trigger an Unstake event, allowing stakers to spend tokens without penalty
+    /// @dev This is expected to be called from the spending contract, which must be first permissioned through a direct call to approve(...)
+    function spend(uint256 amount_, address staker_, address receiver_) external {
+        if(staker_ != msg.sender) {
+            _user_to_approved_spend[staker_][msg.sender] -= amount_;
+        }
+
+        uint256 _balance = _unstake(amount_, msg.sender, receiver_);
```

Comment:
@figs999 the second argument of this function should be staker_? Since this is called from spending contract msg.sender would be contract address I believe.